### PR TITLE
Gemma4

### PR DESCRIPTION
**Pull request by knyazer**

### CODE REVIEW

File: lalamo/model_import/model_configs/huggingface/gemma4.py [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+from lalamo.modules.token_mixers.attention import AttentionConfig
+from lalamo.modules.transformer_layer import TransformerLayerConfig
+
+from .common import HuggingFaceLMConfig, QuantizationConfigType
+
+__all__ = ["HFGemma4Config"]
+
+
+def _round_to_bfloat16(x: float) -> float:
+    return jnp.asarray(x).astype(jnp.bfloat16).item()
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/gemma4.py]: remove function, either inline, or better yet - not convert

File: lalamo/model_import/model_configs/huggingface/gemma4.py [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+
+
+@dataclass(frozen=True)
+class Gemma4RoPEParameters:
+    rope_theta: float
+    rope_type: Literal["default", "proportional"]
+    partial_rotary_factor: float | None = None
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/gemma4.py:42]: do we really need this thingy? does it have values in the config? also remove defaults

File: lalamo/model_import/model_configs/huggingface/gemma4.py [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+    rope_parameters: dict[str, Any] = field(default_factory=dict)
+    final_logit_softcapping: float | None = 30.0
+    vocab_size: int = 262144
+    layer_types: list[Literal["sliding_attention", "full_attention"]] = field(default_factory=list)
+    hidden_activation: str = "gelu_pytorch_tanh"
+    hidden_size_per_layer_input: int = 0
+    vocab_size_per_layer_input: int = 262144
+    num_kv_shared_layers: int = 0
+    use_double_wide_mlp: bool = False
+    tie_word_embeddings: bool = True
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/gemma4.py]: remove all the defaults

File: lalamo/model_import/model_configs/huggingface/gemma4.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+    @property
+    def _sliding_window_sizes(self) -> list[int | None]:
+        return [self.sliding_window if lt == "sliding_attention" else None for lt in self.layer_types]
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/gemma4.py]: inline

File: lalamo/model_import/model_configs/huggingface/gemma4.py [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+
+    def _head_dim_for_layer(self, layer_idx: int) -> int:
+        if self.layer_types[layer_idx] == "full_attention":
+            return self.global_head_dim
+        return self.head_dim
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/gemma4.py]: inline

File: lalamo/model_import/model_configs/huggingface/gemma4.py [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+            if prev_types[j] == target_type:
+                return j
+        return None
+
+    def _intermediate_size_for_layer(self, layer_idx: int) -> int:
+        first_kv_shared = self.num_hidden_layers - self.num_kv_shared_layers
+        is_shared = layer_idx >= first_kv_shared > 0
+        if self.use_double_wide_mlp and is_shared:
+            return self.intermediate_size * 2
+        return self.intermediate_size
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/gemma4.py]: inline, why is it based off the first_kv_shared index? this is not meaningful, i think there should be some nicer indication within the model

File: lalamo/model_import/model_configs/huggingface/gemma4.py [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+
+    def _rope_config_for_type(self, layer_type: str) -> Gemma4RoPEParameters:
+        params = self.rope_parameters.get(layer_type, {})
+        return Gemma4RoPEParameters(
+            rope_theta=params.get("rope_theta", 10000.0),
+            rope_type=params.get("rope_type", "default"),
+            partial_rotary_factor=params.get("partial_rotary_factor"),
+        )
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/gemma4.py]: just remove this thing, either inline or something else, this is bad

File: lalamo/model_import/model_configs/huggingface/gemma4.py [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+
+        # PLE layer config
+        ple_layer_config = None
+        if self.hidden_size_per_layer_input > 0:
+            ple_layer_config = PLELayerConfig(
+                linear_config=linear_config,
+                norm_config=rms_norm_config,
+                ple_dim=self.hidden_size_per_layer_input,
+                activation=GELU(),
+            )
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/gemma4.py]: isn't this always guaranteed to be not None?

File: lalamo/model_import/model_configs/huggingface/gemma4.py [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+            local_rope_config=local_rope_config,
+            layer_configs=tuple(layer_configs),
+            output_norm_config=rms_norm_config,
+            model_dim=self.hidden_size,
+            hidden_dim=self.intermediate_size,
+            context_length=context_length or self.max_position_embeddings,
+        )
+
+        ple_model_config = None
+        if self.hidden_size_per_layer_input > 0:
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/gemma4.py]: same here, no need for the condition

File: lalamo/model_import/model_configs/huggingface/gemma4.py [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+            ple_model_config = PLEModelConfig(
+                ple_dim=self.hidden_size_per_layer_input,
+                num_layers=self.num_hidden_layers,
+                ple_vocab_size=self.vocab_size_per_layer_input,
+                ple_embed_scale=self.hidden_size_per_layer_input**0.5,
+                model_projection_scale=self.hidden_size**-0.5,
+                input_scale=2.0**-0.5,
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/gemma4.py]: can you actually give names to these non-trivial **0.5 or **-0.5 values?

File: lalamo/model_import/model_configs/huggingface/gemma4.py [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+    boi_token_id: int | None = None
+    eoi_token_id: int | None = None
+    video_token_id: int | None = None
+    eoa_token_id: int | None = None
+    eoa_token_index: int | None = None
+    boa_token_id: int | None = None
+    vision_soft_tokens_per_image: int | None = None
+    initializer_range: float = 0.02
+    transformers_version: str = ""
+
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/gemma4.py]: remove defaults

File: lalamo/model_import/model_configs/huggingface/gemma4.py [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+        # Determine the base path for language model weights
+        if any(key.startswith("model.language_model.") for key in weights_dict):
+            base = ParameterPath("model") / "language_model"
+        elif any(key.startswith("language_model.") for key in weights_dict):
+            base = ParameterPath("language_model")
+        else:
+            base = ParameterPath("model")
+
+        # Load PLE token embedding
```
Comment (knyazer) [inline lalamo/model_import/model_configs/huggingface/gemma4.py]: remove useless comments

File: lalamo/modules/token_mixers/attention.py [hunk 13]
```diff
@@ -25,6 +25,14 @@
...
 
 
+def _rms_normalize(
+    x: Float[Array, "*batch channels"],
+    eps: float,
+) -> Float[Array, "*batch channels"]:
+    variance = jnp.mean(jnp.square(x.astype(jnp.float32)), axis=-1, keepdims=True) + eps
+    return (x * jax.lax.rsqrt(variance)).astype(x.dtype)
+
+
```
Comment (knyazer) [inline lalamo/modules/token_mixers/attention.py:35]: can't you just have it as a separate normalization layer instead of adding one more parameters? not sure tho, lmk your thoughts

File: lalamo/modules/token_mixers/attention.py [hunk 14]
```diff
@@ -99,6 +107,8 @@ class AttentionConfig(TokenMixerConfigBase):
     use_rope: bool = True
     # Per-head rotary dimension; if set smaller than head_dim; RoPE is applied to the start of the embedding
     partial_rope_dim: int | None = None
+    # Scale-free RMS normalization on values (used by Gemma 4)
```
Comment (knyazer) [inline lalamo/modules/token_mixers/attention.py]: remove gemma 4 mention

File: lalamo/modules/decoder.py [hunk 15]
```diff
@@ -112,17 +129,39 @@ def empty(self) -> "Decoder":
...
+            ple_projection_norm=ple_projection_norm,
         )
 
 
 class Decoder(LalamoModule[DecoderConfig]):
     embedding: EmbeddingBase
     transformer: Transformer
+    ple_token_embedding: Float[Array, "vocab ple_total_dim"] | None = None
+    ple_model_projection: LinearBase | None = None
+    ple_projection_norm: Normalization | None = None
```
Comment (knyazer) [inline lalamo/modules/decoder.py]: can't you have a special PerLayerEmbedding module that would incorporate all this stuff?

File: lalamo/modules/rope.py [hunk 16]
```diff
@@ -94,13 +94,23 @@ def init(
...
-        embeddings = jnp.concatenate((outer_inverse_frequencies, outer_inverse_frequencies), axis=-1)
+        if rotary_dim is not None and rotary_dim < head_dim:
+            # Pad the rotary embeddings to full head_dim.
+            # Structure: [active, identity, active, identity] so that
+            # rotate_half correctly pairs active dims across the two halves.
+            half_pad = jnp.zeros((num_timesteps, head_dim // 2 - rotary_dim // 2), dtype=jnp.float32)
+            padded_half = jnp.concatenate((outer_inverse_frequencies, half_pad), axis=-1)
+            embeddings = jnp.concatenate((padded_half, padded_half), axis=-1)
+        else:
+            embeddings = jnp.concatenate((outer_inverse_frequencies, outer_inverse_frequencies), axis=-1)
```
Comment (knyazer) [inline lalamo/modules/rope.py]: didn't we already had partial rope somewhere?

File: lalamo/modules/transformer_layer.py [hunk 17]
```diff
@@ -86,19 +96,47 @@ class TransformerLayerConfig:
...
+    def _init_ple(
+        self, model_dim: int, *, empty: bool, key: PRNGKeyArray | None = None
+    ) -> tuple[
+        LinearBase | None,
+        LinearBase | None,
+        Normalization | None,
+    ]:
+        if self.ple_config is None:
+            return None, None, None
+        cfg = self.ple_config
```
Comment (knyazer) [inline lalamo/modules/transformer_layer.py]: don't save letters - just use self.ple_config everywhere, that's more readable

Comment (knyazer) [inline lalamo/modules/transformer_layer.py]: also can't this be a method on actual PLELayer module or something

File: lalamo/modules/transformer_layer.py [hunk 18]
```diff
@@ -170,6 +224,10 @@ class TransformerLayer(LalamoModule[TransformerLayerConfig]):
     pre_mlp_norm: Normalization | None
     mlp: MLPBase
     post_mlp_norm: Normalization | None
+    ple_gate: LinearBase | None = None
+    ple_projection: LinearBase | None = None
+    ple_norm: Normalization | None = None
+    layer_scalar: Float[Array, "1"] | None = None
```
Comment (knyazer) [inline lalamo/modules/transformer_layer.py]: avoid defaults please, this should be a module too i think

File: lalamo/modules/transformer.py [hunk 19]
```diff
@@ -64,42 +64,90 @@ class TransformerConfig:
...
+                PositionalEmbeddingSelector.LOCAL
+                if mixer.sliding_window_size is not None
+                else PositionalEmbeddingSelector.GLOBAL
+            )
+            result.setdefault(selector, mixer.head_dim)
+        return result
+
+    def _init_ropes(self) -> tuple[RoPE | None, RoPE | None]:
+        rope_dims = self._rope_dims_by_selector()
+        head_dims = self._head_dims_by_selector()
```
Comment (knyazer) [inline lalamo/modules/transformer.py]: bad, this should be part of the gemma4 model definition, not overriden in here :(

File: lalamo/modules/decoder.py [hunk 20]
```diff
@@ -194,15 +287,23 @@ def init_static_state(self, batch_size: int, capacity: int) -> State:
         return self.transformer.init_static_state(batch_size, capacity)
 
     def export_weights(self) -> ParameterTree:
-        return dict(
+        result: dict[str, ParameterTree | Array] = dict(
```
Comment (knyazer) [inline lalamo/modules/decoder.py]: not a big fan of this annotation

File: lalamo/modules/rope.py [hunk 21]
```diff
@@ -94,13 +94,23 @@ def init(
...
+        inverse_frequencies = self._scale_inverse_frequencies(inverse_frequencies, freq_dim, self.max_sequence_length)
         outer_inverse_frequencies = jnp.outer(timesteps, inverse_frequencies)
-        embeddings = jnp.concatenate((outer_inverse_frequencies, outer_inverse_frequencies), axis=-1)
+        if rotary_dim is not None and rotary_dim < head_dim:
+            # Pad the rotary embeddings to full head_dim.
+            # Structure: [active, identity, active, identity] so that
+            # rotate_half correctly pairs active dims across the two halves.
+            half_pad = jnp.zeros((num_timesteps, head_dim // 2 - rotary_dim // 2), dtype=jnp.float32)
+            padded_half = jnp.concatenate((outer_inverse_frequencies, half_pad), axis=-1)
+            embeddings = jnp.concatenate((padded_half, padded_half), axis=-1)
```
Comment (knyazer) [inline lalamo/modules/rope.py]: shouldn't this be somehow more explicitly guarded for gemma4? i don't like that gemma4 pr touches rope that is used by almost every other model

File: lalamo/modules/transformer.py [hunk 22]
```diff
@@ -214,19 +195,32 @@ def __call__(
...
+                per_layer_input=per_layer_input,
             )
+
             inner_features = layer_result.outputs
             layer_results.append(layer_result)
-            updated_state_layers.append(layer_result.updated_state)
+
+            if kv_source is not None:
+                # Shared layer: keep the source's state (discard the extended one)
+                updated_state_layers.append(updated_state_layers[kv_source])
```
Comment (knyazer) [inline lalamo/modules/transformer.py]: we don't actually discard it here; maybe we can just drop it, without spending memory on copying it?

File: lalamo/modules/transformer_layer.py [hunk 23]
```diff
@@ -284,7 +386,7 @@ def init_static_state(self, batch_size: int, capacity: int) -> StaticKVCacheLaye
         )
 
     def export_weights(self) -> ParameterTree:
-        result = dict(
+        result: dict[str, ParameterTree | Array] = dict(
```
Comment (knyazer) [inline lalamo/modules/transformer_layer.py:381]: not a big fan of this annotation, just require_mapping(blah) would be better?

File: tests/tracer/test_huggingface_models.py [hunk 24]
```diff
@@ -6,6 +6,7 @@
 from tests.tracer.tracer_huggingface import HFDecoderTracer, ModernBertTracer
 
 MODEL_LIST = [
+    ModelTestSpec("google/gemma-4-E2B-it", DType.FLOAT32, num_tokens=8, token_stride=1),
```
Comment (knyazer) [inline tests/tracer/test_huggingface_models.py]: why num_tokens and token_stride? what's the point

**Review by uuuvn — APPROVED** [review]
